include/zephyr/drivers/location.h

```diff
@@ -0,0 +1,319 @@
+	/** The uptime when the altitude was determined in ticks */
+	k_ticks_t uptime_ticks;
+	/** Accuracy in millimeters */
+	uint32_t accuracy;
+};
+
+/**
+ * @typedef location_position_get_t
+ * @brief API for getting newest position data
+ *
+ * See struct location_position for description of position data
+ */
+typedef int (*location_position_get_t)(const struct device *dev,
+			struct location_position *position);
+
+/**
+ * @typedef location_bearing_get_t
+ * @brief API for getting newest bearing data
+ *
+ * See struct location_bearing for description of bearing data
+ */
+typedef int (*location_bearing_get_t)(const struct device *dev,
+			struct location_bearing *bearing);
+
+/**
+ * @typedef location_speed_get_t
+ * @brief API for getting newest speed data
+ *
+ * See struct location_speed for description of speed data
+ */
+typedef int (*location_speed_get_t)(const struct device *dev,
+			struct location_speed *speed);
+
+/**
+ * @typedef location_altitude_get_t
+ * @brief API for getting newest altitude data
+ *
+ * See struct location_altitude for description of altitude data
+ */
+typedef int (*location_altitude_get_t)(const struct device *dev,
+			struct location_altitude *altitude);
+
+/**
+ * @brief Location events contained in location_events_t
+ */
+#define LOCATION_EVENT_POSITION_UPDATED BIT(0)
+#define LOCATION_EVENT_BEARING_UPDATED BIT(1)
+#define LOCATION_EVENT_SPEED_UPDATED BIT(2)
+#define LOCATION_EVENT_ALTITUDE_UPDATED BIT(3)
+
+/**
+ * @typedef location_events_t
+ * @brief Contains location events represented by single bits
+ */
+typedef uint8_t location_events_t;
+
+/**
+ * @typedef location_event_handler_t
+ * @brief Callback API invoked when a location event is raised
+ */
+typedef void (*location_event_handler_t)(const struct device *dev,
+			location_events_t events, void *user_data);
+
+/**
+ * @typedef location_event_handler_set_t
+ * @brief API for setting the location event handler
+ */
+typedef int (*location_event_handler_set_t)(const struct device *dev,
+			location_event_handler_t handler, void *user_data);
+
+/**
+ * @brief Location API
+ */
+__subsystem struct location_driver_api {
+	location_position_get_t position_get;
+	location_bearing_get_t bearing_get;
+	location_speed_get_t speed_get;
+	location_altitude_get_t altitude_get;
+	location_event_handler_set_t event_handler_set;
+};
+
+/**
+ * @brief Get the latest position from the device instance
+ *
+ * @param dev Device instance
+ * @param position Destination for position data
+ * @return -EINVAL if any argument is NULL
+ * @return -ENOSYS if api call is not implemented by driver
+ * @return -ENOENT if no data is available
+ * @return 0 if successful, negative errno code otherwise
+ */
+__syscall int location_position_get(const struct device *dev, struct location_position *position);
+
+static inline int z_impl_location_position_get(const struct device *dev,
+			struct location_position *position)
+{
+	if (dev == NULL || position == NULL) {
```

Comment:
you don't have to check dev. Up to caller to make sure dev is properly initialized. see other api definitions.

Also, why not using a local "api" variable, just to make it nicer and avoid all the cast calls?

Reply:
I will update this in the location api PR